lib/P6doc/CMD.pm6

```diff
@@ -27,110 +31,158 @@ package P6doc::CMD {
-		my @paths = search-paths() X~ <Type/ Language/>;
-		my @modules;
-		for @paths -> $dir {
-			for dir($dir).sort -> $file {
-				@modules.push: $file.basename.subst( '.'~$file.extension,'') if $file.IO.f;
-			}
-		}
-		@modules.append: list-installed().map(*.name);
-		.say for @modules.unique.sort;
-	}
-
-	multi sub MAIN(Str $file where $file.IO.e, Bool :$n) {
-		say get-docs($file.IO);
-	}
-
-	# index related
-	multi sub MAIN('env') {
-		my Str $index-info = "Index file: {INDEX}";
-		my Str $doc-info = "Doc folder: {get-doc-locations}";
-
-		say $index-info;
-		say $doc-info;
-	}
-
-	multi sub MAIN('build') {
-		say "Building index file...";
-		build-index(INDEX);
-	}
+    multi sub MAIN($docee, Bool :$n) {
+        # On windows, if input is not surrounded by '', it will be malformed
+        # Example: `p6doc X::IO` will pass `X:/:IO` to MAIN.
+        # This should be checked for and corrected.
+
+        return MAIN($docee, :f, :$n) if defined $docee.index('.');
+
+        say get-docs(locate-module($docee).IO, :package($docee));
+    }
+
+    multi sub MAIN($docee, Bool :$f!, Bool :$n) {
+        my ($package, $method) = $docee.split('.');
+        if ! $method {
+
+            if not INDEX.IO.e {
+                say "building index on first run. Please wait...";
+                build-index(INDEX);
+            }
+
+            my %data = from-json slurp(INDEX);
+
+            my $final-docee = disambiguate-f-search($docee, %data);
+
+            # NOTE: This is a temporary fix, since disambiguate-f-search
+            #       does not properly handle independent routines right now.
+            if $final-docee eq '' {
+                $final-docee = ('independent-routines', $docee).join('.');
+            }
+
+            ($package, $method) = $final-docee.split('.');
+
+            my $m = locate-module($package);
+
+            say get-docs($m.IO, :section($method), :$package);
+        } else {
+            my $m = locate-module($package);
+
+            say get-docs($m.IO, :section($method), :$package);
+        }
+    }
+
+    multi sub MAIN(Bool :$l!) {
+        my @paths = search-paths() X~ <Type/ Language/>;
+        my @modules;
+        for @paths -> $dir {
+            for dir($dir).sort -> $file {
+                @modules.push: $file.basename.subst( '.'~$file.extension,'') if $file.IO.f;
+            }
+        }
+        @modules.append: list-installed().map(*.name);
+        .say for @modules.unique.sort;
+    }
+
+    multi sub MAIN(Str $file where $file.IO.e, Bool :$n) {
+        say get-docs($file.IO);
+    }
+
+    # index related
+    multi sub MAIN('env') {
+        my Str $index-info = "Index file: {INDEX}";
+        my Str $doc-info = "Doc folder: {get-doc-locations}";
+
+        say $index-info;
+        say $doc-info;
+    }
+
+    multi sub MAIN('build') {
+        say "Building index file...";
+        build-index(INDEX);
+    }
+    ]
+
+    ###
+    ###
+    ###
+
+    sub USAGE() {
```

Comment:
The `USAGE` message should be clear about commands *and options*. Arguments starting with `-` and `--` are generally known as options. If you don't define your own `USAGE`, that's how the named arguments of `MAIN` will be treated too. I think it's important to keep that here as well.

Furthermore, according to the `MAIN` definitions I also spot long form options of `-r` and `-d`, which I would highly recommend to include. I'm also missing the `-h` in this output, which is accepted by a `MAIN` definition.

Lastly, I generally prefer it when options and commands are sorted alphabetically in the documentation/usage output. This makes manual searching easier.